### ISSUE TITLE
Inconsistent Range behavior when dealing with reverse ranges

### ISSUE THREAD

**Issue by isomarcte** [typelevel/cats-collections#564]
I've noticed some inconsistencies with how several methods on `Range` and that use `Range` are defined when dealing with ranges which go different directions.

### Range.contains ###

`contains` for `Range` is not symmetric around `reverse`.

```scala
scala> val a = Range(1, 10)
a: cats.collections.Range[Int] = Range(1,10)

scala> val b = a.reverse
b: cats.collections.Range[Int] = Range(10,1)

scala> a.contains(b)
res0: Boolean = true

scala> b.contains(a)
res1: Boolean = false
```

Perhaps most troubling,

```scala
scala> Range(1, 10).contains(Range(15, 9))
res0: Boolean = true
```

This is made more confusing when looking at how `Range` is used with `Diet`.

```scala
scala> val a = Range(1, 10)
a: cats.collections.Range[Int] = Range(1,10)

scala> val b = a.reverse
b: cats.collections.Range[Int] = Range(10,1)

scala> Diet.fromRange(b) === Diet.fromRange(a)
res0: Boolean = false

scala> Diet.empty.addRange(b) === Diet.fromRange(a)
res1: Boolean = true
```

There are a few ways to remedy this that come to mind, though I think the most simple would be to change `Range` to be defined as always going in one direction. If a `Range` is desired that is going in the inverse order, then an newtype (perhaps [Inverse](https://github.com/typelevel/cats/pull/4345)) could be used.

**Comment by satorg** [comment]
I would say it is by design: `Range` defines boundaries but not specify `Order` upon creation. Then every single method has to provide `Order` implicitly. In particular, both the `contains` methods take an implicit `Order` parameter. Therefore, if we reverse `Range` then the `Order` used for its methods has also be reversed:
```
scala> val rng1 = Range(1, 10)
rng1: cats.collections.Range[Int] = Range(1,10)

scala> rng1.contains(5)
res0: Boolean = true

scala> val rng2 = rng1.reverse
rng2: cats.collections.Range[Int] = Range(10,1)

scala> rng2.contains(5) // won't work since it uses a regular Order for Int
res1: Boolean = false

scala> rng2.contains(5)(Order.reverse(Order[Int]))
res2: Boolean = true
```

That said, I personally find such an approach controversial quite a bit.

**Comment by satorg** [comment]
> this is one of the things I think we should consider changing for 1.0

If we could change that, then perhaps it would be a cool change. Would we consider making `Range` to be not "all-inclusive" only? I mean, if we could create a `Range` instance having one or both boundaries that would be exclusive, then it would make this type way more useful.

**Comment by armanbilge** [comment]
See also some discussion in https://github.com/typelevel/cats-collections/issues/315#issuecomment-770266231.

**Comment by armanbilge** [comment]
> I would expect to use newtypes (ideally opaque types) to index alternative implementations.

👍 yes, this is the only safe way.

**Comment by isomarcte** [comment]
:+1: You scared me for a minute. 

So coming back to `Range`. If we define ranges as intervals going in a single direction, then I think we can make the interface consistent and relatively easy to use. This would require the use of a newtype for reversing direction. I'm happy to POC that if there are no objections. It would also make the case for moving [Inverse](https://github.com/typelevel/cats/pull/4345) to cats-collections more compelling.

My intended strategy would probably be to introduce a new class `Interval` and deprecate `Range`.

Thoughts?

**Comment by armanbilge** [comment]
Thanks for that distinction, at least for me I think that helped clarify things.

I think (1) is an "interval": it is a set of things, in some natural order. Whereas (2) is a "range" since you are not only specifying the set, but in which order to iterate it. I am not sure if that means we have to reverse the `Order` instance: the direction is a property of the "range", and not of its element. I think :)

**Comment by satorg** [comment]
>  I am not sure if that means we have to reverse the Order instance: the direction is a property of the "range", and not of its element.

In theory yes, it should not be the property of Range. But the way it is currently implemented is that `Order` is used to correlate the Range's boundaries with other items of the same type (including boundaries of another Range). Therefore, currently `Order` magically becomes a property of the Range itself. And yeah, I agree it should not be so...

**Comment by isomarcte** [comment]
@satorg RE 1.CONS

Assuming the inclusion of `Inverse` from the aforementioned PR, do you think something along these lines would work?

```scala
val a: Interval[Int] = Interval(1, 10)
val b: Interval[Inverse[Int]] = a.reverse

def reverse: Interval[Inverse[A]] = 
  Interval(Inverse(end), Inverse(start))
def contains(value: Interval[A]): Boolean = ???
def containsInverse(value: Interval[Inverse[A]]): Boolean = 
  contains(value.reverse)

// Just a rough sketch. I might be able to do something interesting so that
// (a: Interval[Inverse[A]]).reverse => Interval[A], and not
// Interval[Inverse[Inverse[A]]]
:t reverse 
Interval[A] => Interval[Inverse[A]]
:t contains 
Interval[A] => Interval[A] => Boolean
:t containsInverse
Interval[A] => Interval[Inverse[A]] => Boolean
```

And I'm completely fine making it exclusive on the upper bound. I think that would even be required, since if we require that `start <= end`, then the only want to define the empty interval is to have the upper bound be exclusive.

**Comment by isomarcte** [comment]
I can also just make a POC PR as well, if that would be easier for discussion.

**Comment by satorg** [comment]
> And I'm completely fine making it exclusive on the upper bound. I think that would even be required, since if we require that start <= end, then the only want to define the empty interval is to have the upper bound be exclusive.

Yes, I think it makes sense as an option... But on the other hand.... Imagine that we need to make `Interval` for the entire `Short` type range. In that case `Interval(Short.MinVal, Short.MaxVal)` wouldn't work since the last value will be always excluded.

Another (but similar) example:
```scala
sealed trait MyNum

object MyNum {
  object One extends MyNum
  object Two extends MyNum
  object Three extends MyNum
  object Four extends MyNum
}
```
Apparently we can define some `Order` for such an enum as well as `Discrete` (or perhaps even `BoundedEnumerable` from **cats**) for that. But if `Interval` can only have an exclusive upper bound, then there's no way to define such an `Interval` that would contain all the members of an enum like one above.

**Comment by satorg** [comment]
As a bottomline: there's a controverity: if `Interval` only allows both lower and upper bounds to be inclusive, then there's no way to define an empty `Interval`. Otherwise, if it always defined as [inclusive, exclusive), then we cannot describe a full range of some limited set of values.

This is why I think it should allow both options.

**Comment by satorg** [comment]
On the second thought... Perhaps, the idea is too drastic, but... is there really a necessity for having classes like `Range` or `Interval` in principle, taking into account that there are cool typeclasses like `UnboundedEnumerable`, `LowerBoundedEnumerable`, `BoundedEnumerable`, etc already defined in **cats**?

**Comment by isomarcte** [comment]
@satorg I would say yes. An `Interval` type, not only providing some utility itself, is fundamental to a class of data structures [called Interval Trees](https://en.wikipedia.org/wiki/Interval_tree), of which we have a variant in cats-collection already, e.g. `Diet`.

Having a base type with the correct semantics to implement these is, I think, quite useful.

I am of course heavily biased here. I started down this path because a nice `Interval` type and an Interval Tree which supports key value pairs would be amazing for my work in [idna4s](https://github.com/typelevel/idna4s). :wink:

**Comment by satorg** [comment]
I basically agree. But what I am thinking about (and yeah, perhaps it is too crazy) is to make `Diet` more generic and don't stick to a particular interval/range type. Therefore if we made implementation of `Diet` based on typeclasses rather than a particular range implementation, then it should be possible to plug any range implementation into it, including `scala.collection.immutable.Range`, something like
```scala
class Diet[A, R] {}
```
where `A` – element type and `R` – range type
